File: Pigeonhole_Sort/Pigeonhole_Sort.java
import java.util.Scanner;

class Pigeonhole_Sort {
    // Method for getting the maximum value
    public static int getMax(int[] inputArray) {
        int maxValue = inputArray[0];
        for (int i = 1; i < inputArray.length; i++) {
            if (inputArray[i] > maxValue) {
                maxValue = inputArray[i];
            }
        }
        return maxValue;
    }

    // Method for getting the minimum value
    public static int getMin(int[] inputArray) {
        int minValue = inputArray[0];
        for (int i = 1; i < inputArray.length; i++) {
            if (inputArray[i] < minValue) {
                minValue = inputArray[i];
            }
        }
        return minValue;
    }

    public static void sort(int[] input) {
        //calculate size for array( no of pigeon holes)
        int max_no = getMax(input);
        int min_no = getMin(input);
        int size = max_no - min_no + 1;

        //create holes
        int[] holes = new int[size];
        for (int i = 0; i < input.length; i++)
            holes[input[i] - min_no]++;

        int index = 0;

        for (int j = 0; j < size; j++)
            while (holes[j]-- > 0)
                input[index++] = j + min_no;
    }

    public static void main(String[] args) 
    {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter size of array");
        int n = sc.nextInt();  
        int[] array = new int[n];
        System.out.println("Enter elements of array");
        for(int i = 0; i < n; i++)
            array[i] = sc.nextInt();
        sort(array);
        for(int i = 0; i < n; i++)
             System.out.println(array[i] + " ");
    }
}

/*
Input:
n = 4
array = [1, 4, 32, 2]

Output 
array = [1, 2, 4, 32]
*/

File: Pigeonhole_Sort/Pigeonhole_Sort.kt
import java.util.*

fun main(args: Array<String>) {

    var n: Int = Integer.valueOf(readLine())    
    var arr = arrayOfNulls<Int>(n)              
    var read = Scanner(System.`in`)          
    for(i in 0..(n-1)){
        arr[i] = read.nextInt()                 
    }   
    var min = arr[0]                            
    var max = arr[0]                            
    //finding the minimum and maximum elements in the array
    for(i in 0..(n-1)){
        if(arr[i]!! < min!!)
            min = arr[i]
        if(arr[i]!! > max!!)
            max = arr[i]
    }

    var noOfPigeonHoles = max!! - min!! + 1     //size of the pigeonhole array
    var pigeonHoles = IntArray(noOfPigeonHoles) //pigeonhole array
    Arrays.fill(pigeonHoles,0)                  //initializing all the elements to 0

    //pigeonhole sort algorithm
    for(i in 0..(n-1)){
        var x = arr[i]!! - min                  
        pigeonHoles[x]++
    }
    var index = 0
    
    for(i in 0..(noOfPigeonHoles-1)){
        while(pigeonHoles[i] > 0){
            arr[index++] = i + min
            pigeonHoles[i]--
        }
    }
    
    //printing the array
    println("Sorted Array")
    for(i in 0..(n-1)){
        println(arr[i])
    }
}

/*
Sample Input
5
1 5 4 2 0
Sample Output
Sorted Array
0 
1 
2 
4 
5
*/
